dietpi/dietpi-software

```diff
@@ -10639,6 +10649,54 @@ _EOF_
 			[[ -f '/mnt/dietpi_userdata/gitea/custom/conf/app.ini' ]] && G_CONFIG_INJECT 'RUN_USER[[:blank:]]' 'RUN_USER = gitea' /mnt/dietpi_userdata/gitea/custom/conf/app.ini
 		fi
 
+		if To_Install 177 forgejo # Forgejo
+		then
+			# ARMv7: Fork of gitea, do not support ARMv7 - https://github.com/go-gitea/gitea/issues/6700
+			case $G_HW_ARCH in
+				3) local arch='arm64';;
+				10) local arch='amd64';;
+				*) local arch='arm-6';;
+			esac
+
+			local fallback_url="https://codeberg.org/forgejo/forgejo/releases/download/v7.0.2/forgejo-7.0.2-linux-$arch.xz"
+			Download_Install "$(curl -sSfL 'https://codeberg.org/api/v1/repos/forgejo/forgejo/releases/latest' | jq '.' | mawk -F\" "/\"browser_download_url\": \".*\/forgejo-[^\"\/]*-linux-$arch\.xz\"/{print \$4}")" /mnt/dietpi_userdata/forgejo/forgejo
```

Comment:
As an alternative to `jg`, `mawk` could loop through all fields, until it finds one which matches the URL pattern. I can write that, if needed.

Reply:
I am not familiar with awk, but I found a manual, I a, looking into it.  https://www.gnu.org/software/gawk/manual/gawk.html

I could not use the gitea awk version because the json from github is readable https://api.github.com/repos/go-gitea/gitea/releases/latest, while the codeberg/forgejo api returns a single string. https://codeberg.org/api/v1/repos/forgejo/forgejo/releases/latest

https://codeberg.org/api/swagger#/repository/repoGetLatestRelease

I can find the fields I need with
`curl -sSfL 'https://codeberg.org/api/v1/repos/forgejo/forgejo/releases/latest' | mawk -F: -v RS=, '$1~/"browser_download_url"/'` 
but there are some tweaks to do.

Reply:
Got it

`curl -sSfL 'https://codeberg.org/api/v1/repos/forgejo/forgejo/releases/latest' | mawk -F:\" -v RS=, '/"browser_download_url":".*\/forgejo-[^\"\/]*-linux-arm64\.xz"/{gsub("\"|}| ]","",$2);print $2}'`

I will update the file and test it, please let me know if I can improve the regex.

Reply:
It seems to be working now. 

I also added a line to start Forgejo after the installation is done.  `systemctl status` was showing it as inactive before.

Reply:
With the "loop through fields" I actually meant something like this:
```sh
curl -sSfL 'https://codeberg.org/api/v1/repos/forgejo/forgejo/releases/latest' | mawk -F\" "{for(i=NF;i>0;i--) if(\$i~/^https:.*linux-$arch\.xz$/) print \$i}"
```
But changing the record delimiter isn't a bad idea either. Running some tests, it seems to perform slightly faster. Can be simplified a little:
```sh
curl -sSfL 'https://codeberg.org/api/v1/repos/forgejo/forgejo/releases/latest' | mawk -F\" -v RS=, "/^\"browser_download_url\":\".*-linux-$arch\.xz\"/{print \$4}"
```
Since every `key: value` pair must be separated by comma, no need to worry about additional `"`. And no need to separate fields by `"` as well as `:`, then removing all `"` to effectively use `:` as field separator only. Just use `"` only 😉.

Reply:
That is brilliant, it also solves the problem with the closing brackets.
It is also faster, I guess removing the `gsub` and adding the `exit` statement helps with that.

Thanks for the guidance.

I tested it here and updated the code.